### PR TITLE
update user role filter to use nested query

### PR DESCRIPTION
**Pull request by ylwu-amzn**
*Issue #, if available:*

*Description of changes:*
We changed AD index mapping to use nested type for user field (#291 ). This PR is to update user role filter to nested query.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by ohltyler — APPROVED** [review]

**Review by yizheliu-amazon — APPROVED** [review]
thanks for the change

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/util/ParseUtils.java [hunk 1]
```diff
@@ -390,4 +397,41 @@ public static SearchSourceBuilder generateEntityColdStartQuery(
...
+        String userFieldName = "user";
+        String userBackendRoleFieldName = "user.backend_roles.keyword";
+        if (user == null) {
+            // For old monitor and detector, they have no user field, user = null
+            ExistsQueryBuilder userRolesFilterQuery = QueryBuilders.existsQuery(userFieldName);
+            NestedQueryBuilder nestedQueryBuilder = new NestedQueryBuilder(userFieldName, userRolesFilterQuery, ScoreMode.None);
+            boolQueryBuilder.mustNot(nestedQueryBuilder);
+        } else if (user.getBackendRoles() == null || user.getBackendRoles().size() == 0) {
+            // For simple FGAC user, they may have no backend roles, these users should be able to see detectors
+            // of other users whose backend role is empty. user != null, user.backend_role == null
```
Comment (kaituo) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/util/ParseUtils.java:412]: In this case, why don't we filter based on users' roles or just user name?

Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/util/ParseUtils.java:412]: Good question.
1.Why not use user's roles?
It's ok if all users use our predefined roles. But user can create custom role with arbitrary name. It's impossible to guess what permission the user role covers. Maybe role "abc" has exactly the same permission with "123".

2.Why not use user name ?
The main use case is for enterprise org/team. For one team, team members should have same permission and they can access team's resources no matter it's created by themselves or other team members. If limit resource access to only the creator, they have to share one account across team which is not good.

Comment (kaituo) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/util/ParseUtils.java:412]: 1. does filtering using backend role has the same issue?  backend role "abc" has the same permission with "123"?

2. user name based solution is opposite to your current solution: it is overly restrictive while your current solution is overly permissive.

Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/util/ParseUtils.java:412]: 1.Backend role has no mapping to permission. They just map to individuals or orgs. So if user1 and usr2 has same backend role, they are in same org.

2.Agree.

**Pull request merged**